slack_importer: Add cross-linking notification to first thread messages.

PieterCK opened the pull request:
Currently we send all thread messages to their own thread topic. This updates our Slack importer to send the first thread message to the main topic and add a cross-linking notification message to the first thread message, linking it to its thread topic.

Previously part of: #30166
Fixes part of: #27626
CZO: [#user questions > What to do with Slack imported topics?](https://chat.zulip.org/#narrow/channel/138-user-questions/topic/What.20to.20do.20with.20Slack.20imported.20topics.3F)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
(Updated Nov 28)
<img width="1090" height="533" alt="image" src="https://github.com/user-attachments/assets/b3a447b6-03a0-4299-bcf5-d9798ca94634" />
<img width="1090" height="605" alt="image" src="https://github.com/user-attachments/assets/5b444c0b-b806-4394-a9b9-1b7edbeb4886" />

the original thread message is now in the main "imported from Slack" topic instead of in its branched-off thread topic. 
| Slack | Zulip |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/57c55491-a07e-4fb0-ac8d-de01aa28da24) | ![image](https://github.com/user-attachments/assets/ab3b8ad6-8050-4d80-b167-3170f64dd673) |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
slack_importer: Add cross-linking message for Slack thread.

This commit sends the original thread message to the main import topic
instead of dumping it in together with its replies in a separate thread
topic.

It also appends a cross-linking notification message to the original
thread message, linking it to its thread topic.

Fixes part of #27626.

PieterCK commented on zerver/data_import/slack.py:
```diff
@@ -1097,15 +1109,37 @@ def channel_message_to_zerver_message(
+                # The first thread message has a `thread_ts` that matches its
+                # `message_ts`. We want to send this message to the main import
+                # channel and add a cross-linking notification message to the
+                # thread topic.
```
For the first thread message, It looks like the `thread_ts` value is copied from the messages `message_ts` value once it has a reply and thus turned into a thread. So `thread_ts` is actually not when the thread is created, but when the first thread message is sent.

PieterCK commented on zerver/tests/test_slack_importer.py:
```diff
@@ -1557,48 +1566,55 @@ def test_channel_message_to_zerver_message_with_threads(
         self.assertEqual(zerver_message[3]["content"], expected_thread_2_message_1_content)
-        self.assertEqual(zerver_message[3][EXPORT_TOPIC_NAME], expected_thread_2_topic_name)
-        # Record that truncation should use the full maximum topic length.
-        self.assert_length(zerver_message[3][EXPORT_TOPIC_NAME], 60)
```
first thread message is sent to the main topic, so we update the truncation check to use a thread reply instead.

PieterCK commented:
@mateuszmandera -- when you have time, could you please review this? Thanks

PieterCK committed:
test_slack_importer: Re-focus Slack thread conversion tests.

This re-organizes several tests to reduce busy work when making
adjascent changes.

mateuszmandera commented:
Looks good to me! I think this is a nice enough standalone change that we can merge without blocking on also doing the `Notification Bot` message; and can leave that for a follow-up?

@alya Probably worth a look from you to check if the string etc. look good, and if my above assumption is correct.

mateuszmandera commented:
@aman Can you integration review?

amanagr commented on zerver/data_import/slack.py:
```diff
@@ -1109,25 +1189,17 @@ def channel_message_to_zerver_message(
-        if is_direct_message_type:
-            topic_name = ""
+        topic_name = create_topic_name_for_message(
+            channel_name=channel_name,
```
Oops! We are using `channel_name` without defining it for every case. Let's add a test case for that too.

PieterCK replied:
Ah, good catch, updated the code to:
```diff
diff --git a/zerver/data_import/slack.py b/zerver/data_import/slack.py
index 70c0c4f316..aa2c4d82d5 100644
--- a/zerver/data_import/slack.py
+++ b/zerver/data_import/slack.py
@@ -1024,7 +1024,7 @@ def is_slack_thread_message(convert_slack_threads: bool, message: ZerverFieldsT)


 def create_topic_name_for_message(
-    channel_name: str,
+    channel_name: str | None,
     content: str,
     convert_slack_threads: bool,
     is_direct_message_type: bool,
@@ -1037,6 +1037,8 @@ def create_topic_name_for_message(
     if is_direct_message_type:
         return ""

+    assert channel_name is not None
+
     # Slack's unthreaded messages go into a single topic, while
     # threaded messages are put into separate thread topics.
     if not is_slack_thread_message(convert_slack_threads, message):
@@ -1135,6 +1137,7 @@ def channel_message_to_zerver_message(
             continue
         rendered_content = None

+        channel_name: str | None = None
         if "channel_name" in message:
             is_direct_message_type = False
             recipient_id = slack_recipient_name_to_zulip_recipient_id[message["channel_name"]]
```

> Let's add a test case for that too.

And in `test_channel_message_to_zerver_message` we have a couple existing asserts that test this:
```python
        # Test that topic_name is set to '\x07' for direct messages and
        # group direct messages.
        self.assertEqual(zerver_message[6][EXPORT_TOPIC_NAME], Message.DM_TOPIC)
        self.assertEqual(zerver_message[8][EXPORT_TOPIC_NAME], Message.DM_TOPIC)
```

amanagr commented:
>  @'aman' Can you integration review?

You tagged the wrong person. :)

PieterCK committed:
test_slack_importer: Add a missing field in test fixture.

Message payloads that we pass to `channel_message_to_zerver_message`
have been updated by `get_messages_iterator` to include additional
fields such as "channel_name".

amanagr commented:
LGTM! Thanks!
